### ISSUE TITLE
workflow message become 'runtime error: invalid memory address or nil pointer dereference'

### ISSUE THREAD

**Issue by qingfeng777** [argoproj/argo-workflows#7662]
## Summary

What happened/what you expected to happen?
some of workflows's  message become 'runtime error: invalid memory address or nil pointer dereference'.
i don't know how to reproduce it .

`
Name:                b9114cdb-872b-4599-81d6-77f116f281af
Namespace:           default
ServiceAccount:      default
Status:              Error
Message:             runtime error: invalid memory address or nil pointer dereference
Conditions:          
 PodRunning          False
 Completed           True
Created:             Wed Jan 26 04:24:51 +0800 (7 hours ago)
Started:             Wed Jan 26 04:24:51 +0800 (7 hours ago)
Finished:            Wed Jan 26 06:38:23 +0800 (4 hours ago)
Duration:            2 hours 13 minutes
Progress:            4108/4108
ResourcesDuration:   1062h40m52s*(1 cpu),38464h35m4s*(100Mi memory)

STEP                                        TEMPLATE           PODNAME                                          DURATION  MESSAGE

`

What version of Argo Workflows are you running?
v3.2.6

## Diagnostics

What Kubernetes provider are you using?
Amazon EKS
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.11-eks-f17b81", GitCommit:"f17b810c9e5a82200d28b6210b458497ddfcf31b", GitTreeState:"clean", BuildDate:"2021-10-15T21:46:21Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by qingfeng777** [comment]
i checked the code, only one place maybe change wf's Phase and Message. at the /workflow/controller/operator.go  ：
https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L2002
so,i think the bug is in this func, https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L3316:

`func (woc *wfOperationCtx) deletePDBResource(ctx context.Context) error {
	if woc.execWf.Spec.PodDisruptionBudget == nil {
		return nil
	}
	err := waitutil.Backoff(retry.DefaultRetry, func() (bool, error) {
              **# maybe here woc.execWf  should be used other than woc.wf ?**  case i see wf maybe set to be nil at other place.
		err := woc.controller.kubeclientset.PolicyV1beta1().PodDisruptionBudgets(woc.wf.Namespace).Delete(ctx, woc.wf.Name, metav1.DeleteOptions{})
		if apierr.IsNotFound(err) {
			return true, nil
		}
		return !errorsutil.IsTransientErr(err), err
	})
	if err != nil {
		// TODO ,if have more early log, can search those.
		woc.log.WithField("err", err).Error("Unable to delete PDB resource for workflow.")
		return err
	}
	woc.log.Infof("Deleted PDB resource for workflow.")
	return nil
}`

**Comment by sarabala1979** [comment]
Can you provide the reproducible sample workflow?
Have you enable the PDB on your workflow?

**Comment by avifried1** [comment]
happening to us as well, pretty complex wf so no way for me to produce a distilled sample for reproducibility. But should point out that these workflows also appear to then be stuck on `running` (although they failed) which also affects our concurrency lock, which is not released until we restart the controller.

**Comment by qingfeng777** [comment]
> Have you enable the PDB on your workflow?



> Can you provide the reproducible sample workflow? Have you enable the PDB on your workflow?
> […](#)
> Sent from my iPhone
> On Jan 27, 2022, at 1:59 AM, 名白 ***@***.***> wrote: ﻿ i checked the code, only one place maybe change wf's Phase and Message. at the /workflow/controller/operator.go ： https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L2002 so,i think the bug is in this func, https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L3316: func (woc *wfOperationCtx) deletePDBResource(ctx context.Context) error { if woc.execWf.Spec.PodDisruptionBudget == nil { return nil } err := waitutil.Backoff(retry.DefaultRetry, func() (bool, error) { **# maybe here woc.execWf should be used other than woc.wf ?** case i see wf maybe set to be nil at other place. err := woc.controller.kubeclientset.PolicyV1beta1().PodDisruptionBudgets(woc.wf.Namespace).Delete(ctx, woc.wf.Name, metav1.DeleteOptions{}) if apierr.IsNotFound(err) { return true, nil } return !errorsutil.IsTransientErr(err), err }) if err != nil { // TODO ,if have more early log, can search those. woc.log.WithField("err", err).Error("Unable to delete PDB resource for workflow.") return err } woc.log.Infof("Deleted PDB resource for workflow.") return nil } — Reply to this email directly, view it on GitHub, or unsubscribe. Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you are subscribed to this thread.

**Comment by qingfeng777** [comment]
> Can you provide the reproducible sample workflow? Have you enable the PDB on your workflow?
i don't know how to reproduce it.
PDB on workflow is disable.

**Comment by alexec** [comment]
Can you please attach the workflow controller's logs?

**Comment by qingfeng777** [comment]
next time it happens

**Comment by qingfeng777** [comment]
it happened again at today 01:51, but i can't get logs from the wf-ctrl pod,  i can only get logs  after 02:10， like laste time ,
the wf-ctrl pod have not restarted,  i don't know why.  
i will set log collector, and try  again when it happens again.

**Comment by qingfeng777** [comment]
```
time="2022-02-08T04:01:16.496Z" level=error msg="Recovered from panic" namespace=default r="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 347 [running]:\nruntime/debug.Stack(0xc037fe702e, 0x1f22520, 0x331a940)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2(0xc02b228e40, 0x24f2c28, 0xc000114000)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:192 +0xd1\npanic(0x1f22520, 0x331a940)\n\t/usr/local/go/src/runtime/panic.go:971 +0x499\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).processAggregateNodeOutputs(0xc02b228e40, 0xc0c0564fa0, 0xc120380180, 0x16, 0xc04e57a000, 0xdd, 0x117, 0x0, 0x0)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2696 +0xc2b\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeSteps(0xc02b228e40, 0x24f2c28, 0xc000114000, 0xc08be21c20, 0x24, 0xc04e72af40, 0xc0a73530b0, 0x2a, 0xc21415b440, 0x24dd0f0, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/steps.go:133 +0x12fb\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc02b228e40, 0x24f2c28, 0xc000114000, 0xc08be21c20, 0x24, 0x24dd0f0, 0xc0435d12c0, 0xc04e72aec0, 0x0, 0x0, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1800 +0x258e\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc02b228e40, 0x24f2c28, 0xc000114000)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:344 +0xf3e\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc000519400, 0x24f2c28, 0xc000114000, 0x0)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:770 +0x945\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000519400)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:692 +0x9b\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0c74536e0)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155 +0x5f\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0c74536e0, 0x24ab6c0, 0xc02d019d70, 0x1, 0xc1050ece40)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156 +0x9b\nk8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0c74536e0, 0x3b9aca00, 0x0, 0x1, 0xc1050ece40)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133 +0x98\nk8s.io/apimachinery/pkg/util/wait.Until(0xc0c74536e0, 0x3b9aca00, 0xc1050ece40)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90 +0x4d\ncreated by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:309 +0x48a\n" workflow=b14c50e6-b9a2-40a9-bc44-6067a70644bc
time="2022-02-08T04:01:16.496Z" level=info msg="Updated message  -> runtime error: invalid memory address or nil pointer dereference" namespace=default workflow=b14c50e6-b9a2-40a9-bc44-6067a70644bc
time="2022-02-08T04:02:04.876Z" level=info msg=healthz age=5m0s err="<nil>" instanceID= labelSelector="!workflows.argoproj.io/phase,!workflows.argoproj.io/controller-instanceid" managedNamespace=
time="2022-02-08T04:02:58.051Z" level=error msg="Recovered from panic" namespace=default r="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 360 [running]:\nruntime/debug.Stack(0xc082a20d5e, 0x1f22520, 0x331a940)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2(0xc03d075980, 0x24f2c28, 0xc000114000)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:192 +0xd1\npanic(0x1f22520, 0x331a940)\n\t/usr/local/go/src/runtime/panic.go:971 +0x499\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).processAggregateNodeOutputs(0xc03d075980, 0xc0de96f2a0, 0xc06cf637d0, 0x16, 0xc0b9100000, 0x385, 0x45d, 0x0, 0x0)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2696 +0xc2b\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeSteps(0xc03d075980, 0x24f2c28, 0xc000114000, 0xc08c354d50, 0x24, 0xc110f22cc0, 0xc067abb170, 0x2a, 0xc06d797b00, 0x24dd0f0, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/steps.go:133 +0x12fb\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc03d075980, 0x24f2c28, 0xc000114000, 0xc08c354d50, 0x24, 0x24dd0f0, 0xc03d075ec0, 0xc110f22c40, 0x0, 0x0, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1800 +0x258e\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc03d075980, 0x24f2c28, 0xc000114000)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:344 +0xf3e\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc000519400, 0x24f2c28, 0xc000114000, 0x0)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:770 +0x945\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000519400)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:692 +0x9b\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0c7453d50)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155 +0x5f\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0c7453d50, 0x24ab6c0, 0xc107ae5bf0, 0xc048f9de01, 0xc1050ece40)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156 +0x9b\nk8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0c7453d50, 0x3b9aca00, 0x0, 0x1, 0xc1050ece40)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133 +0x98\nk8s.io/apimachinery/pkg/util/wait.Until(0xc0c7453d50, 0x3b9aca00, 0xc1050ece40)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90 +0x4d\ncreated by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:309 +0x48a\n" workflow=df97ba9d-7092-4a5c-9690-3af2d6455baa
time="2022-02-08T04:02:58.052Z" level=info msg="Updated message  -> runtime error: invalid memory address or nil pointer dereference" namespace=default workflow=df97ba9d-7092-4a5c-9690-3af2d6455baa
```

**Comment by qingfeng777** [comment]
```
time="2022-02-08T04:02:58.051Z" level=error msg="Recovered from panic" namespace=default 

r="runtime error: invalid memory address or nil pointer dereference" 

stack="
goroutine 360 [running]:
runtime/debug.Stack(0xc082a20d5e, 0x1f22520, 0x331a940)
/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2)(0xc03d075980, 0x24f2c28, 0xc000114000)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/operator.go:192](http://github.com/argoproj/argo-workflows/workflow/controller/operator.go:192) +0xd1
panic(0x1f22520, 0x331a940)
/usr/local/go/src/runtime/panic.go:971 +0x499
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).processAggregateNodeOutputs](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).processAggregateNodeOutputs)(0xc03d075980, 0xc0de96f2a0, 0xc06cf637d0, 0x16, 0xc0b9100000, 0x385, 0x45d, 0x0, 0x0)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/operator.go:2696](http://github.com/argoproj/argo-workflows/workflow/controller/operator.go:2696) +0xc2b
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeSteps](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeSteps)(0xc03d075980, 0x24f2c28, 0xc000114000, 0xc08c354d50, 0x24, 0xc110f22cc0, 0xc067abb170, 0x2a, 0xc06d797b00, 0x24dd0f0, ...)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/steps.go:133](http://github.com/argoproj/argo-workflows/workflow/controller/steps.go:133) +0x12fb
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate)(0xc03d075980, 0x24f2c28, 0xc000114000, 0xc08c354d50, 0x24, 0x24dd0f0, 0xc03d075ec0, 0xc110f22c40, 0x0, 0x0, ...)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/operator.go:1800](http://github.com/argoproj/argo-workflows/workflow/controller/operator.go:1800) +0x258e
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate)(0xc03d075980, 0x24f2c28, 0xc000114000)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/operator.go:344](http://github.com/argoproj/argo-workflows/workflow/controller/operator.go:344) +0xf3e
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem)(0xc000519400, 0x24f2c28, 0xc000114000, 0x0)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/controller.go:770](http://github.com/argoproj/argo-workflows/workflow/controller/controller.go:770) +0x945
[github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000519400)](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000519400))
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/controller.go:692](http://github.com/argoproj/argo-workflows/workflow/controller/controller.go:692) +0x9b
[k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0c7453d50)](http://k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0c7453d50))
/go/pkg/mod/[k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155](http://k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155) +0x5f
[k8s.io/apimachinery/pkg/util/wait.BackoffUntil](http://k8s.io/apimachinery/pkg/util/wait.BackoffUntil)(0xc0c7453d50, 0x24ab6c0, 0xc107ae5bf0, 0xc048f9de01, 0xc1050ece40)
/go/pkg/mod/[k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156](http://k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156) +0x9b
[k8s.io/apimachinery/pkg/util/wait.JitterUntil](http://k8s.io/apimachinery/pkg/util/wait.JitterUntil)(0xc0c7453d50, 0x3b9aca00, 0x0, 0x1, 0xc1050ece40)
/go/pkg/mod/[k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133](http://k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133) +0x98
[k8s.io/apimachinery/pkg/util/wait.Until](http://k8s.io/apimachinery/pkg/util/wait.Until)(0xc0c7453d50, 0x3b9aca00, 0xc1050ece40)
/go/pkg/mod/[k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90](http://k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90) +0x4d
created by [github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading](http://github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading)
/go/src/[github.com/argoproj/argo-workflows/workflow/controller/controller.go:309](http://github.com/argoproj/argo-workflows/workflow/controller/controller.go:309) +0x48a
" 
workflow=df97ba9d-
```

**Comment by alexec** [comment]
Error is actually from this line:

https://github.com/argoproj/argo-workflows/blob/v3.2.6/workflow/controller/operator.go#L2696

`p.Value` must be nil. This line needs a nil check.

**Comment by alexec** [comment]
https://github.com/argoproj/argo-workflows/pull/7517/files

**Comment by alexec** [comment]
Fixed in v3.2.7
